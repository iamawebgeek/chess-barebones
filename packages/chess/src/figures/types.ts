export enum Figure {
  KING = 'king',
  QUEEN = 'queen',
  ROOK = 'rook',
  BISHOP = 'bishop',
  KNIGHT = 'knight',
  PAWN = 'pawn',
}
