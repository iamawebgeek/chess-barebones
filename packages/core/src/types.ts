export enum Direction {
  NORTH,
  SOUTH,
  WEST,
  EAST,
  NORTH_WEST,
  SOUTH_WEST,
  NORTH_EAST,
  SOUTH_EAST,
}

export const DirectionToVectorMap: Record<Direction, [number, number]> = {
  [Direction.NORTH]: [0, 1],
  [Direction.SOUTH]: [0, -1],
  [Direction.WEST]: [-1, 0],
  [Direction.EAST]: [1, 0],
  [Direction.NORTH_WEST]: [-1, 1],
  [Direction.SOUTH_WEST]: [-1, -1],
  [Direction.NORTH_EAST]: [1, 1],
  [Direction.SOUTH_EAST]: [1, -1],
};

export type Coordinate = {
  x: number;
  y: number;
};

export type DeepReadonly<T> = T extends (infer R)[]
  ? DeepReadonlyArray<R>
  : T extends (...args: unknown[]) => unknown
    ? T
    : T extends object
      ? DeepReadonlyObject<T>
      : T;

export type DeepReadonlyArray<T> = ReadonlyArray<DeepReadonly<T>>;

export type DeepReadonlyObject<T> = {
  readonly [P in keyof T]: DeepReadonly<T[P]>;
};
