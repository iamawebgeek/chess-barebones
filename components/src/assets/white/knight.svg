<svg width="72" height="72" viewBox="0 0 72 72" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M36.8465 4.09469C35.5528 2.58588 33.596 1.81556 31.621 2.03764C29.646 2.25972 27.9091 3.44538 26.9829 5.20381C26.0641 6.94811 25.1639 9.32513 24.7993 11.9253C24.707 12.5835 24.644 13.2959 24.6321 14.0447C19.2955 17.0323 14.6388 22.8426 12.3338 32.0629C12.1722 32.7092 11.9931 33.0376 11.8318 33.2795C11.6048 33.6201 11.3125 33.9505 10.5642 34.7469C9.10833 36.2963 7.07057 38.4658 5.20686 42.5038C2.36265 48.6663 4.87149 54.9269 9.33576 58.1054C11.9629 59.9759 15.2252 60.7929 18.4757 60.276C17.9667 62.4534 17.7903 64.4281 17.8686 65.8485C18.0441 69.029 20.6742 71.5181 23.8595 71.5181H61.1546C64.4683 71.5181 67.1546 68.8318 67.1546 65.5181C67.1546 53.6703 66.3955 41.8629 63.1875 32.4524C60.1115 23.4295 54.4504 15.8217 44.5989 13.1367L36.8465 4.09469ZM10.6546 45.0182C12.1108 41.863 13.6222 40.2545 14.9368 38.8554C16.3305 37.3722 17.5031 36.1242 18.1546 33.5182C19.8374 26.7869 22.8224 22.6703 25.9659 20.3222C27.0254 19.5308 28.1029 18.9402 29.1546 18.5182C29.8086 18.2557 30.654 18.1061 31.5387 18.0273C31.0614 17.199 30.796 16.2871 30.6878 15.3477C30.5626 14.2614 30.6476 13.1382 30.858 12.0647C31.1605 10.5214 31.7222 9.08079 32.2914 8.00006L41.3253 18.5366C56.6005 21.4118 60.5169 38.6267 61.0785 59.5181C61.1315 61.4884 61.1546 63.4914 61.1546 65.5181H23.8595C23.8479 65.3076 23.8472 65.0546 23.8595 64.7655C23.9129 63.5194 24.21 61.6021 24.9323 59.5181C25.401 58.1661 26.0486 56.7439 26.9247 55.3894C28.0707 53.6176 29.6076 51.9614 31.6464 50.7292C31.8494 50.6065 32.0463 50.4814 32.2374 50.3541C38.2198 46.3674 38.4559 40.1853 37.794 37.5182C36.3583 43.4186 33.8166 44.619 30.3108 46.2747C29.9689 46.4362 29.6178 46.602 29.2577 46.7769C27.2338 47.76 24.9244 49.0309 22.3529 51.4386C22.1222 51.6545 21.8895 51.8796 21.6546 52.1145C15.7509 58.0182 7.65462 51.5182 10.6546 45.0182Z" fill="#34364C"/>
</svg>
