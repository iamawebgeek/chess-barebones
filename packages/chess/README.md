# @chess-barebones/chess

[![Version](https://img.shields.io/npm/v/@chess-barebones/chess.svg)](https://npm.im/@chess-barebones/chess)
[![Downloads](https://img.shields.io/npm/dm/@chess-barebones/chess.svg)](https://npmcharts.com/compare/chess-barebones?minimal=true)
[![License](https://img.shields.io/npm/l/@chess-barebones/chess.svg)](https://www.npmjs.com/package/@chess-barebones/chess)

## Installation

To get started, install the `@chess-barebones/chess` package via pnpm, npm or yarn:

```
npm install @chess-barebones/chess
```

## Documentation

Find information on how to use this package on the [documentation website](https://github.com/iamawebgeek/chess-barebones).
