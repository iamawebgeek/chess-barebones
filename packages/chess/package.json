{
  "name": "@chess-barebones/chess",
  "version": "0.1.0",
  "description": "Chess specific additions for @chess-barebones/core",
  "keywords": [
    "chess",
    "chess barebones",
    "chess puzzle",
    "chess backend",
    "checkmate",
    "san",
    "fen"
  ],
  "license": "LGPL-2.1-or-later",
  "type": "module",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./package.json": "./package.json"
  },
  "files": [
    "dist",
    "README.md",
    "LICENSE"
  ],
  "sideEffects": false,
  "scripts": {
    "build": "tsup",
    "test": "vitest run",
    "test:coverage": "vitest run --coverage",
    "lint": "eslint src",
    "typecheck": "tsc --noEmit"
  },
  "author": "iamawebgeek <iamawebgeek@gmail.com>",
  "repository": {
    "type": "git",
    "url": "https://github.com/iamawebgeek/chess-barebones.git"
  },
  "bugs": {
    "url": "https://github.com/iamawebgeek/chess-barebones/issues"
  },
  "homepage": "https://iamawebgeek.github.io/chess-barebones",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@chess-barebones/core": "workspace:^"
  }
}
