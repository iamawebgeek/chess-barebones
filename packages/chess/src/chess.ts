import {
  AbstractBoard,
  combineHandlers,
  Direction,
  Game,
  Handler,
  InvalidStateError,
  Player,
} from '@chess-barebones/core';

import { Figure } from './figures';
import {
  CheckmateProcessor,
  FiftyMovesProcessor,
  InsufficientMaterialsProcessor,
  RepetitionProcessor,
  StalemateProcessor,
} from './processors';
import { AlgebraicNotationSerializer, RegularChessMove } from './serializers';

export enum Color {
  WHITE = 'white',
  BLACK = 'black',
}

export class RegularChess extends Game<Figure, RegularChessMove, Color> {
  public constructor(board: AbstractBoard<Figure>, handler: Handler = {}) {
    const serializer = new AlgebraicNotationSerializer(board);
    super(
      board,
      serializer,
      combineHandlers([
        handler,
        // Do not change
        // The order is important
        new CheckmateProcessor(board),
        new StalemateProcessor(board),
        new InsufficientMaterialsProcessor(board),
        new FiftyMovesProcessor(board),
        new RepetitionProcessor(board),
      ]),
    );
    this.addPlayer(new Player(Color.WHITE, Direction.SOUTH));
    this.addPlayer(new Player(Color.BLACK, Direction.NORTH));
  }

  public start() {
    if (this.state.players.length !== 2) {
      throw new InvalidStateError(
        this,
        'Regular chess can be played with only 2 players, consider making your variation with Game class',
      );
    }
    super.start();
  }

  public getPlayerByColor(color: Color) {
    return this.state.players.find((player) => player.color === color)!;
  }
}
