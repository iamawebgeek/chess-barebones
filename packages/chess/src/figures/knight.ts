import {
  AbstractBoard,
  AbstractFigure,
  Coordinate,
  Player,
} from '@chess-barebones/core';

import { Figure } from './types';

const moves = [
  [2, 1],
  [2, -1],
  [-2, 1],
  [-2, -1],
  [1, 2],
  [1, -2],
  [-1, 2],
  [-1, -2],
] as Readonly<[number, number][]>;

export class Knight extends AbstractFigure<Figure> {
  public constructor(
    owner: Player,
    board: AbstractBoard<Figure>,
    coordinate: Coordinate,
  ) {
    super(Figure.KNIGHT, owner, board, coordinate);
  }

  public getAllMoves(): Coordinate[] {
    const { captured, coordinate } = this.state;
    if (captured) return [];
    return moves
      .map((vector) => this.board.getCoordinateWithVector(coordinate, vector))
      .filter((move) => move !== null)
      .flat();
  }

  public getReach(): Coordinate[] {
    return this.getAllMoves();
  }

  public getAvailableMoves(): Coordinate[] {
    return this.getReach().filter(
      (coordinate) => this.board.getFigure(coordinate)?.owner !== this.owner,
    );
  }
}
