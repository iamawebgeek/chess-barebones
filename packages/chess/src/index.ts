export * from './board';
export * from './figures';
export * from './factory';
export * from './initializers';
export * from './processors';
export * from './decorators';
export * from './serializers';
export * from './chess';
