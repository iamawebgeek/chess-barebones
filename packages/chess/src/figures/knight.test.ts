import {
  Board8x8,
  castFigure,
  Direction,
  FigureFactory,
  NotCaptured,
  Player,
} from '@chess-barebones/core';
import { describe, expect, it } from 'vitest';

import { Knight } from './knight';
import { Figure } from './types';
import { Color } from '../chess';

const player1 = new Player(Color.WHITE, Direction.NORTH);
const player2 = new Player(Color.BLACK, Direction.SOUTH);

const figureMap = {
  [Figure.KNIGHT]: Knight,
} as const;

const figureFactory: FigureFactory<Figure> = {
  create(name, player, board, coordinate) {
    const FigureClass = figureMap[name as keyof typeof figureMap];
    return castFigure<NotCaptured, Figure>(
      new FigureClass(player, board, coordinate),
    );
  },
};

describe('Knight', () => {
  it('moves to given coordinate and captures on landing', () => {
    const board = new Board8x8<Figure>(figureFactory);
    const knight = board.createFigure(Figure.KNIGHT, player1, 'b5');
    knight.move('c7');
    expect(board.getFigure('c7')).toBe(knight);
    expect(board.getFigure('b5')).toBeNull();

    const enemy1 = board.createFigure(Figure.KNIGHT, player2, 'e6');
    knight.move('e6');
    expect(board.getFigure('e6')).toBe(knight);
    expect(board.getFigure('c7')).toBeNull();
    expect(enemy1.state.captured).toBe(true);

    const enemy2 = board.createFigure(Figure.KNIGHT, player2, 'e8');
    knight.move('g7');
    expect(board.getFigure('g7')).toBe(knight);
    knight.move('e8');
    expect(board.getFigure('e8')).toBe(knight);
    expect(enemy2.state.captured).toBe(true);
  });

  it('reaches all diagonals', () => {
    const board = new Board8x8<Figure>(figureFactory);
    const figure = board.createFigure(Figure.KNIGHT, player1, 'b5');
    expect(figure.getReach()).toMatchSnapshot();
    figure.move('a1');
    expect(figure.getReach()).toMatchSnapshot();
  });

  it('returns all moves and available moves correctly with blocking and enemies', () => {
    const board = new Board8x8<Figure>(figureFactory);
    const knight = board.createFigure(Figure.KNIGHT, player1, 'd4');

    board.createFigure(Figure.KNIGHT, player1, 'e6');
    board.createFigure(Figure.KNIGHT, player1, 'c2');
    board.createFigure(Figure.KNIGHT, player2, 'f5');

    const allMoves = knight
      .getAllMoves()
      .map((c) => board.serializeCoordinate(c))
      .sort();
    expect(allMoves).toEqual(
      ['b3', 'b5', 'c2', 'c6', 'e2', 'e6', 'f3', 'f5'].sort(),
    );

    const availableMoves = knight
      .getAvailableMoves()
      .map((c) => board.serializeCoordinate(c))
      .sort();
    expect(availableMoves).toEqual(['b3', 'b5', 'c6', 'e2', 'f3', 'f5'].sort());
  });
});
