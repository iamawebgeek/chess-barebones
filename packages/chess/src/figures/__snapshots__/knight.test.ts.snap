// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Knight > reaches all diagonals 1`] = `
[
  {
    "x": 4,
    "y": 6,
  },
  {
    "x": 4,
    "y": 4,
  },
  {
    "x": 3,
    "y": 7,
  },
  {
    "x": 3,
    "y": 3,
  },
  {
    "x": 1,
    "y": 7,
  },
  {
    "x": 1,
    "y": 3,
  },
]
`;

exports[`Knight > reaches all diagonals 2`] = `
[
  {
    "x": 3,
    "y": 2,
  },
  {
    "x": 2,
    "y": 3,
  },
]
`;
